File: cli/backend/openweathermap/openweathermap_current.py
import time

from weather_core.backend import WeatherCondition, WeatherData, WindData


class OpenWeatherMapCurrent(WeatherData):
    def __new__(cls, *args, **kwargs):
        return super().__new__(cls)

    def __init__(self, data):
        self.time = int(time.time())
        self.temperature = int(data.weather.main.temp)
        self.min_temp = int(data.weather.main.temp_min)
        self.max_temp = int(data.weather.main.temp_max)
        self.feels_like = int(data.weather.main.feels_like)
        self.dewpoint = data.weather.main.humidity
        self.wind = WindData(data.weather.wind.speed, data.weather.wind.deg)
        self.aqi = data.air_quality.list[0].main["aqi"]
        self.cloud_cover = data.weather.clouds.all
        conditions = []
        self.condition_sentence = ""
        self.condition_ids = self.get_condition_ids()

        for condition in data.weather.weather:
            conditions.append(WeatherCondition(condition.id))
        self.conditions = conditions
        self.condition_sentence = self.get_conditions_sentence()

File: src/backend/openweathermap/openweathermap_current.rs
use crate::backend::openweathermap::openweathermap_json::{
    OpenWeatherMapAirQualityJson, OpenWeatherMapJson,
};
use crate::backend::weather_condition::WeatherCondition;
use crate::backend::weather_data::{get_conditions_sentence, WeatherDataRS};
use crate::backend::wind_data::WindData;
use crate::local::weather_file::WeatherFile;
use crate::now;

pub struct OpenWeatherMapCurrent {
    data: OpenWeatherMapJson,
    aqi: OpenWeatherMapAirQualityJson,
}
impl OpenWeatherMapCurrent {
    pub fn new(data: OpenWeatherMapJson, aqi: OpenWeatherMapAirQualityJson) -> Self {
        OpenWeatherMapCurrent { data, aqi }
    }
}
impl WeatherDataRS for OpenWeatherMapCurrent {
    fn get_time(&self) -> i128 {
        now() as i128
    }

    fn get_temperature(&self) -> f32 {
        self.data.main.temp as f32
    }

    fn get_min_temp(&self) -> f32 {
        self.data.main.temp_min as f32
    }

    fn get_max_temp(&self) -> f32 {
        self.data.main.temp_max as f32
    }

    fn get_wind(&self) -> WindData {
        WindData {
            speed: self.data.wind.speed,
            heading: self.data.wind.deg,
        }
    }

    fn get_raw_data(&self) -> String {
        serde_json::to_string_pretty(&self.data).expect("dump to string failed")
    }

    fn get_dewpoint(&self) -> f32 {
        self.data.main.humidity as f32
    }

    fn get_feels_like(&self) -> f32 {
        self.data.main.feels_like as f32
    }

    fn get_aqi(&self) -> u8 {
        self.aqi.list[0]
            .main
            .get("aqi")
            .expect("aqi not found")
            .abs_diff(0)
    }

    fn get_cloud_cover(&self) -> u8 {
        self.data.clouds.all
    }

    fn get_conditions(&self) -> Vec<WeatherCondition> {
        let weather_file = WeatherFile::weather_codes();
        let mut conditions: Vec<WeatherCondition> = Vec::new();
        for condition in self.data.weather.clone() {
            conditions.push(WeatherCondition::new(
                condition.id as u16,
                &weather_file.data,
            ))
        }
        conditions
    }

    fn get_condition_sentence(&self) -> String {
        get_conditions_sentence(self.get_conditions())
    }
}
